br/pkg/metautil/statsfile.go

```diff
@@ -0,0 +1,235 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metautil
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/errors"
+	backuppb "github.com/pingcap/kvproto/pkg/brpb"
+	berrors "github.com/pingcap/tidb/br/pkg/errors"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tidb/br/pkg/utils"
+	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/statistics/handle"
+	statstypes "github.com/pingcap/tidb/pkg/statistics/handle/types"
+	statsutil "github.com/pingcap/tidb/pkg/statistics/handle/util"
+	"golang.org/x/sync/errgroup"
+)
+
+var maxStatsJsonTableSize = 32 * 1024 * 1024 // 32 MiB
+var inlineSize = 8 * 1024                    // 8 KiB
+
+func getStatsFileName(physicalID int64) string {
+	return fmt.Sprintf("backupmeta.schema.stats.%09d", physicalID)
+}
+
+// A lightweight function wrapper to dump the statistic
+type StatsWriter struct {
+	storage storage.ExternalStorage
+	cipher  *backuppb.CipherInfo
+
+	// final stats file indexes
+	statsFileIndexes []*backuppb.StatsFileIndex
+
+	// temporary variables, clear after each flush
+	totalSize int
+	statsFile *backuppb.StatsFile
+}
+
+func newStatsWriter(
+	storage storage.ExternalStorage,
+	cipher *backuppb.CipherInfo,
+) *StatsWriter {
+	return &StatsWriter{
+		storage: storage,
+		cipher:  cipher,
+
+		statsFileIndexes: make([]*backuppb.StatsFileIndex, 0),
+
+		totalSize: 0,
+		statsFile: &backuppb.StatsFile{},
+	}
+}
+
+func (s *StatsWriter) writeStatsFileAndClear(ctx context.Context, physicalID int64) error {
+	fileName := getStatsFileName(physicalID)
+	content, err := proto.Marshal(s.statsFile)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if len(s.statsFileIndexes) == 0 && len(content) < inlineSize {
+		s.statsFileIndexes = append(s.statsFileIndexes, &backuppb.StatsFileIndex{InlineData: content})
+		return nil
+	}
+
+	checksum := sha256.Sum256(content)
+
+	encryptedContent, iv, err := Encrypt(content, s.cipher)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := s.storage.WriteFile(ctx, fileName, encryptedContent); err != nil {
+		return errors.Trace(err)
+	}
+
+	s.statsFileIndexes = append(s.statsFileIndexes, &backuppb.StatsFileIndex{
+		Name:     fileName,
+		Sha256:   checksum[:],
+		SizeEnc:  uint64(len(encryptedContent)),
+		SizeOri:  uint64(len(content)),
+		CipherIv: iv,
+	})
+
+	// clear the temporary variables
+	s.totalSize = 0
+	s.statsFile = &backuppb.StatsFile{}
```

Comment:
NIT: prefer use a reset function here. and new StatsFile with a fixed length blocks.